test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+
+        const a = lhs.data;
+        const b = rhs.data;
+        var c = new Tensor(rank=2,eltType=eltType);
+        c.reshapeDomain({0..#m, 0..#q});
+        forall (i,j) in c.domain with (ref c) {
+            const row = a[i,..];
+            const col = b[..,j];
+            c[i,j] = + reduce (row * col);
+        }
+        return c;
+    }
+
+    operator /(lhs: Tensor(?d,?eltType), c: eltType) {
+        const data = lhs.data / c;
+        return new Tensor(data);
+    }
+    operator -(lhs: Tensor(?d,?eltType), c: eltType) {
+        const data = lhs.data - c;
+        return new Tensor(data);
+    }
+
+    // Sigmoid function
+    proc _sigmoid(x: real): real {
+        return 1.0 / (1.0 + Math.exp(-x));
+    }
+    // Derivative of sigmoid function
+    proc _sigmoidPrime(x: real): real {
+        const s = _sigmoid(x);
+        return s * (1.0 - s);
+    }
+
+    // Apply sigmoid function to each element of tensor
+    proc sigmoid(t: Tensor(?d)): Tensor(d) {
+        return t.fmap(_sigmoid);
+    }
+
+    // Apply derivative of sigmoid function to each element of tensor
+    proc sigmoidPrime(t: Tensor(?d)): Tensor(d) {
+        return t.fmap(_sigmoidPrime);
+    }
+
+    // Get forbenius norm before sqrt
+    proc frobeniusNormPowTwo(t: Tensor(?d)): real {
+        const AA = t.data ** 2.0;
+        return + reduce AA;
+    }
+
+    // Apply exponential function to each element of tensor
+    proc exp(t: Tensor(?d)): Tensor(d) {
+        var y = new Tensor(t.domain);
+        foreach i in t.domain do
+            y.data[i] = Math.exp(t.data[i]);
+        return y;
+    }
+
+    // Wikipedia implementation (helper for randn)
+    // mu : mean
+    // sigma : standard deviation
+    proc boxMuller(mu: real, sigma: real) {
+        var u1 = rng.getNext();
+        var u2 = rng.getNext();
+        var z0 = sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.pi * u2);
+        return mu + (sigma * z0);
+    }
+    proc normal() {
+        return boxMuller(0.0,1.0);
+    }
+
+    // Initialize a tensor with random values from a normal distribution
+    proc randn(shape: int ...?d): Tensor(d,real) {
+        var t = new Tensor((...shape));
+        for i in t.domain {
+            t.data[i] = normal();
+        }
+        return t;
+    }
+
+    // Initialize a tensor with random values from a normal distribution
+    proc randn(shape: int ...?d, mu: real, sigma: real): Tensor(d,real) {
+        var t = new Tensor((...shape));
+        var m: [t.data.domain] real;
+        for i in m.domain {
+            var x: real = boxMuller(mu,sigma);
+            m[i] = x;
+        }
+        return new Tensor(m);
+    }
+
+    // Initialize a tensor with zero values
+    proc zeros(shape: int ...?d): Tensor(d,real) {
+        return new Tensor((...shape));
+    }
+
+    // Shuffle a tensor in place
+    proc shuffle(ref x) {
+        Random.shuffle(x,seed=(rng.getNext() * 10):int);
+    }
```

Comment:
This might make more sense as a method.

Reply:
Yes, you are right

Reply:
Oh actually, this is just for arrays. This is so that it will be deterministic with respect to `rng`.